File: 05-sandbox-execution/golang/providers/openai.go
package providers

import (
	"context"
	"errors"
	"io"
	"strings"

	openai "github.com/sashabaranov/go-openai"
	"os"
)

type openAIProvider struct {
	client *openai.Client
	model  string
}

func NewOpenAIProvider() Provider {
	config := openai.DefaultConfig(os.Getenv("OPENAI_API_KEY"))
	if base := os.Getenv("OPENAI_API_BASE_URL"); base != "" {
		config.BaseURL = base
	}
	model := os.Getenv("OPENAI_MODEL")
	if model == "" {
		model = "gpt-4o"
	}
	return &openAIProvider{client: openai.NewClientWithConfig(config), model: model}
}

func (p *openAIProvider) Name() string       { return "openai" }
func (p *openAIProvider) ContextWindow() int { return 128_000 }

func (p *openAIProvider) Chat(ctx context.Context, messages []Message) (string, error) {
	msgs := make([]openai.ChatCompletionMessage, len(messages))
	for i, m := range messages {
		msgs[i] = openai.ChatCompletionMessage{Role: m.Role, Content: m.Content}
	}
	resp, err := p.client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model:    p.model,
		Messages: msgs,
	})
	if err != nil {
		return "", err
	}
	return resp.Choices[0].Message.Content, nil
}

func (p *openAIProvider) Stream(ctx context.Context, messages []Message, onToken func(string)) (string, error) {
	msgs := make([]openai.ChatCompletionMessage, len(messages))
	for i, m := range messages {
		msgs[i] = openai.ChatCompletionMessage{Role: m.Role, Content: m.Content}
	}
	stream, err := p.client.CreateChatCompletionStream(ctx, openai.ChatCompletionRequest{
		Model:    p.model,
		Messages: msgs,
	})
	if err != nil {
		return "", err
	}
	defer stream.Close()

	var sb strings.Builder
	for {
		resp, err := stream.Recv()
		if errors.Is(err, io.EOF) {
			break
		}
		if err != nil {
			return "", err
		}
		if len(resp.Choices) > 0 {
			token := resp.Choices[0].Delta.Content
			if token != "" {
				onToken(token)
				sb.WriteString(token)
			}
		}
	}
	return sb.String(), nil
}

File: 09-multi-agent/nodejs/src/providers/openai.ts
import OpenAI from 'openai';
import type { Message, ContentBlock, Provider } from './types.ts';

// Convert our ContentBlock[] to the OpenAI SDK's content format.
function toOpenAIContent(content: string | ContentBlock[]): string | OpenAI.ChatCompletionContentPart[] {
  if (typeof content === 'string') return content;
  return content.map((block) => {
    if (block.type === 'text') return { type: 'text' as const, text: block.text };
    return {
      type: 'image_url' as const,
      image_url: { url: `data:${block.source.media_type};base64,${block.source.data}` },
    };
  });
}

export function createOpenAIProvider(): Provider {
  const client = new OpenAI({
    apiKey: process.env['OPENAI_API_KEY'],
    baseURL: process.env['OPENAI_API_BASE_URL'],
  });
  const model = process.env['OPENAI_MODEL'] ?? 'gpt-4o';

  return {
    name: 'openai',
    contextWindow: 128_000,

    async chat(messages: Message[]): Promise<string> {
      const completion = await client.chat.completions.create({
        model,
        messages: messages.map(m => ({ role: m.role, content: toOpenAIContent(m.content) })),
      });
      return completion.choices[0].message.content ?? '';
    },

    async stream(messages: Message[], onToken: (token: string) => void): Promise<string> {
      const stream = await client.chat.completions.create({
        model,
        messages: messages.map(m => ({ role: m.role, content: toOpenAIContent(m.content) })),
        stream: true,
      });
      let full = '';
      for await (const chunk of stream) {
        const token = chunk.choices[0]?.delta?.content ?? '';
        if (token) { onToken(token); full += token; }
      }
      return full;
    },
  };
}
